src/Function.cpp

```diff
@@ -71,12 +71,70 @@ using namespace std;
 
 ///////////////////////////////////////////////////////////////////////////////
 
+static vector<FunctionAttribute*> attributes;
 static vector<Function*> FuncList;		// List of all functions in the program
 static vector<FactMgr*>  FMList;        // list of fact managers for each function
 static long cur_func_idx;				// Index into FuncList that we are currently working on
 static bool param_first=true;			// Flag to track output of commas
 static int builtin_functions_cnt;
 
+FunctionAttribute::FunctionAttribute(string name, int prob)
+	:  attribute(name), attribute_probability(prob)
+{
+	//Nothing else to do
+}
+
+void
+FunctionAttribute::OutputAttribute(std::ostream &out, string option)
+{
+	out << " __attribute__((" << attribute << option << "))";
+}
+
+BooleanFunctionAttribute::BooleanFunctionAttribute(string name, int prob)
+	: FunctionAttribute(name, prob)
+{
+	//Nothing else to do
+}
+
+void
+BooleanFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, "");
+}
+
+MultiValuedFunctionAttribute::MultiValuedFunctionAttribute(string name, int prob, vector<string> arguments)
+	: FunctionAttribute(name, prob), attribute_values(arguments)
+{
+	//Nothing else to do
+}
+
+void
+MultiValuedFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, "(\"" +attribute_values[rnd_upto(attribute_values.size())] + "\")");
+}
+
+void
+Function::GenerateAttributes()
+{
+	if(CGOptions::func_attr_flag()){
+		vector<string> common_func_attributes = {"artificial", "flatten", "no_reorder", "hot", "cold", "noipa", "nothrow", "deprecated", "no_icf", "no_profile_instrument_function", "no_instrument_function", "no_sanitize_address", "no_sanitize_thread", "no_sanitize_undefined", "no_split_stack", "noinline", "noplt", "stack_protect"};
+		vector<string>::iterator itr;
+		for(itr = common_func_attributes.begin(); itr < common_func_attributes.end(); itr++)
+			attributes.push_back(new BooleanFunctionAttribute(*itr, FuncAttrProb));
+
+		if(rnd_flipcoin(FuncAttrProb))
+			attributes.push_back(new BooleanFunctionAttribute("used", FuncAttrProb));
+		else
+			attributes.push_back(new BooleanFunctionAttribute("unused", FuncAttrProb));
+
+		attributes.push_back(new MultiValuedFunctionAttribute("visibility", FuncAttrProb, {"default", "hidden", "protected", "internal"}));
+		attributes.push_back(new MultiValuedFunctionAttribute("no_sanitize", FuncAttrProb, {"no_sanitize_address", "no_sanitize_thread", "no_sanitize_undefined"}));
```

Comment:
The values of the attribute are wrong, please take a look at `man gcc` and search for `-fsanitize=`.

Reply:
Got it. Should I push recent suggested changes?

Reply:
Push it right after you'll fix all review notes. Not before please.

Reply:
All the review notes have been fixed except providing default value to `name`. It is not required right now because `used` and `unused` are moved to `BooleanFunctionAttribute`

Reply:
Good, then feel free to push the branch. Thanks.

Reply:
Done. Please check it.